File: src/main/java/org/mitallast/queue/raft/resource/ResourceRegistry.java
package org.mitallast.queue.raft.resource;

import javaslang.collection.HashMap;
import javaslang.collection.Map;
import javaslang.collection.Vector;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.mitallast.queue.common.codec.Message;
import org.mitallast.queue.raft.protocol.RaftSnapshot;
import org.mitallast.queue.raft.protocol.RaftSnapshotMetadata;

public class ResourceRegistry {
    private final static Logger logger = LogManager.getLogger();

    private volatile Vector<ResourceFSM> resources = Vector.empty();
    private volatile Map<Class, ResourceHandler> handlers = HashMap.empty();

    public synchronized void register(ResourceFSM fsm) {
        resources = resources.append(fsm);
    }

    public synchronized <T extends Message> void register(Class<T> type, ResourceHandler<T> handler) {
        handlers = handlers.put(type, handler);
    }

    @SuppressWarnings("unchecked")
    public Message apply(long index, Message event) {
        ResourceHandler handler = handlers.getOrElse(event.getClass(), null);
        if (handler != null) {
            return handler.apply(index, event);
        } else {
            logger.warn("resource handler not found: {}", event);
            return null;
        }
    }

    public RaftSnapshot prepareSnapshot(RaftSnapshotMetadata snapshotMeta) {
        Vector<Message> snapshots = resources.flatMap(r -> r.prepareSnapshot(snapshotMeta));
        return new RaftSnapshot(snapshotMeta, snapshots);
    }

    @FunctionalInterface
    public interface ResourceHandler<T extends Message> {
        Message apply(long index, T event);
    }
}

File: src/main/java/org/mitallast/queue/raft/resource/ResourceRegistry.kt
package org.mitallast.queue.raft.resource

import io.vavr.collection.HashMap
import io.vavr.collection.Map
import io.vavr.collection.Vector
import io.vavr.control.Option
import org.mitallast.queue.common.codec.Message
import org.mitallast.queue.raft.protocol.RaftSnapshot
import org.mitallast.queue.raft.protocol.RaftSnapshotMetadata


@Suppress("UNCHECKED_CAST")
class ResourceRegistry {
    @Volatile private var resources = Vector.empty<ResourceFSM>()
    @Volatile private var handlers: Map<Class<*>, (Long, event: Message) -> Option<Message>> = HashMap.empty()

    @Synchronized
    fun register(fsm: ResourceFSM) {
        resources = resources.append(fsm)
    }

    @Synchronized
    fun <T : Message> register(type: Class<T>, handler: (Long, event: T) -> Option<Message>) {
        handlers = handlers.put(type, handler as (Long, event: Message) -> Option<Message>)
    }

    fun <T : Message> apply(index: Long, event: T): Option<Message> {
        return handlers.get(event.javaClass)
            .map { it as (Long, event: T) -> Option<Message> }
            .flatMap { it.invoke(index, event) }
    }

    fun prepareSnapshot(snapshotMeta: RaftSnapshotMetadata): RaftSnapshot {
        val snapshots = resources.flatMap { r -> r.prepareSnapshot(snapshotMeta) }
        return RaftSnapshot(snapshotMeta, snapshots)
    }
}
